### PR TITLE
Large mint/burn Agent

### PR DESCRIPTION
**Pull request by yasmine9**
# Large Mints - Burns on PGL contract

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Agent
- [ ] Fix an agent error
- [ ] Update an agent logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Agent PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the agent properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The agent README is updated
- [x] The agent README sufficiently describes what the agent does
- [x] All the alerts that the agent emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Agent contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The code does what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the agent is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [ ] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The agent follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [ ] Agent can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)

## Comments (optional)

If there is some useful information or important note for reviewing/understanding this agent please describe it here.

### CODE REVIEW

File: Benqi-Agents/Large-Mint-Burn/package.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    "push": "forta-agent push",
+    "disable": "forta-agent disable",
+    "enable": "forta-agent enable",
+    "keyfile": "forta-agent keyfile",
+    "test": "jest"
+  },
+  "dependencies": {
+    "forta-agent": "^0.0.39"
+  },
+  "devDependencies": {
```
Comment (Kalzak) [inline Benqi-Agents/Large-Mint-Burn/package.json:24]: `prettier` seems to be missing, please add it and run `npm run format`.

**Review by Kalzak — CHANGES_REQUESTED** [review]

File: Benqi-Agents/Large-Mint-Burn/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  - Type is always set to "Info"
+  - Metadata includes:
+    - `source`: address of the user initializing the Burn.
+    - `amount0`: amount of `QI` tokens that were Burned.
+    - `amount1`: amount of `WAVAX` tokens that were Burned.
+    - `to`: address to which the tokens were transfered.
+
+## Test Data
+
+There's no txs including large Mints/Burns on PGL contract. We have deployed a PoC contract to emit the monitored events. The agent behaviour can be verified with the following test transactions (Ropsten):
```
Comment (Kalzak) [inline Benqi-Agents/Large-Mint-Burn/README.md:38]:
```suggestion
There are no transactions including large Mints/Burns on PGL contract. We have deployed a PoC contract to emit the monitored events. The agent behaviour can be verified with the following test transactions (Ropsten):
```

File: Benqi-Agents/Large-Mint-Burn/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+> note that `TESTNET_PGL_CONTRACT` should be used instead of `PGL_CONTRACT`.
```
Comment (Kalzak) [inline Benqi-Agents/Large-Mint-Burn/README.md:39]: We should mention where and how this should be changed in-case there are some people using this agent who aren't too familiar with Forta.

File: Benqi-Agents/Large-Mint-Burn/README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+- 0x212324e7e0909810cbf002414488a7d448a219f5a7e06d824e7b43925287b8bd (`Mint` with regulat amounts - no Finding generated).
```
Comment (Kalzak) [inline Benqi-Agents/Large-Mint-Burn/README.md:41]:
```suggestion
- 0x212324e7e0909810cbf002414488a7d448a219f5a7e06d824e7b43925287b8bd (`Mint` with regular amounts - no Finding generated).
```

File: Benqi-Agents/Large-Mint-Burn/README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+- 0xdf20b418ca234228ef4390ccd7d5d391bffaf20209075bfbf0fbe5ba0347fc67 (`Mint` with large QI amount - Finding generated).
+- 0xb4f867b6558c371b19b3f16e43468287f0d91e8b156a0b54ff3f3b957fa25ce4 (`Burn` with large QI-WAVAX amounts- Finding generated).
```
Comment (Kalzak) [inline Benqi-Agents/Large-Mint-Burn/README.md:43]:
```suggestion
- 0xb4f867b6558c371b19b3f16e43468287f0d91e8b156a0b54ff3f3b957fa25ce4 (`Burn` with large QI-WAVAX amounts - Finding generated).
```

File: Benqi-Agents/Large-Mint-Burn/README.md (hunk 3 above)
Comment (Kalzak) [reply Benqi-Agents/Large-Mint-Burn/README.md:39]: Something simple like mentioning to change the inputs to `provideHandleTransaction` in the default exports in `agent.ts` should suffice.

**Review by Kalzak — COMMENTED** [review]
Documentation recommendations

**Comment by 0xtaf** [comment]
.gitignore, .prettierrc and .dockerignore files are missing.

File: Benqi-Agents/Large-Mint-Burn/src/agent.spec.ts [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+  getReserves: mockGetReserves,
+};
+
+// init the agent
+const handler = provideHandleTransaction(TEST_PGL_CONTRACT, 10, mockFetcher as any);
+
+beforeEach(() => {
+  mockFetcher.getReserves.mockClear();
+  mockFetcher.getReserves.mockResolvedValue([BigNumber.from(1000), BigNumber.from(2000)]);
+});
```
Comment (e1Ru1o) [inline Benqi-Agents/Large-Mint-Burn/src/agent.spec.ts:72]: Is this supposed to be here?

File: Benqi-Agents/Large-Mint-Burn/src/agent.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        findings.push(createFinding(log));
+    });
+    return findings;
+  };
+
+export default {
+  handleTransaction: provideHandleTransaction(
+    PGL_CONTRACT,
+    PERCENT,
+    new ReservesFetcher(getEthersProvider(), PGL_CONTRACT)
```
Comment (e1Ru1o) [inline Benqi-Agents/Large-Mint-Burn/src/agent.ts:60]: would be nice to use in the agent logic the same pglAddress used in the Fetcher (i.e `fetcher.pglAddress`) two avoid passing twice the address here

**Review by e1Ru1o — APPROVED** [review]

**Review by RCantu92 — APPROVED** [review]
Just a couple of general comments:

I think if we have a test that generates multiple Findings, it is probably unnecessary to then write a test whose only goal is to confirm multiple Findings are generated. (e.g. `"should return a finding when both QI-WAVAX amounts are above the threshold"` and `should return multiple findings`.)

Secondly, it would probably be better to deploy to the mainnet's own testnet. In this case deploying the PoC contract to the Avalanche testnet rather than Ropsten, which was used. However, this has already been discussed, and I think this will be implemented going forward.

**Review by Kalzak — APPROVED** [review]

**Pull request merged**